lessons/07-mcp/README.md

```diff
@@ -0,0 +1,482 @@
+
+**You**: "This doesn't do much right?"
+
+**Time beetle**: "No, not yet. But we can add resources and tools to it. A tool is something that can be used to perform a specific task, while a resource is something that can be used to provide data or functionality to the tool. Let's add some below:
+
+```typescript
+server.tool("add",
+   { a: z.number(), b: z.number() },
+   async ({ a, b }) => ({
+     content: [{ type: "text", text: String(a + b) }]
+   })
+ );
+```
+
+In the preceding code, we:
+
+- Defined a tool called "add" that takes two numbers as input and returns their sum as output.
+- Described the input parameters and its validation using zod. The `z.number()` function is used to validate the input data, ensuring that both `a` and `b` are numbers.
+- The `async` function is used to perform the addition and return the result in a specific format.
+
+**You**: "Interesting, so the tool is like a function that can be called by the MCP client?"
+
+**Time beetle**: "Exactly! The MCP client can call this tool and pass the required parameters. Let's add a resource to the server as well:
+
+```typescript
+server.resource(
+   "greeting",
+   new ResourceTemplate("greeting://{name}", { list: undefined }),
+   async (uri, { name }) => ({
+     contents: [{
+       uri: uri.href,
+       text: `Hello, ${name}!`
+     }]
+   })
+ );
+```
+
+In this code, we:
+
+- Defined a resource called "greeting" that takes a name as input and returns a greeting message.
+- Used a resource template to define how to call the resource `greeting//{name}`. This schema is used to define the resource's URI format, which includes a placeholder for the name.
+- The `async` function is used to generate the greeting message based on the provided name.
+
+**You**: "So the resource is like a data source, this could be a database, file or even an API? If this was a for a file I would use a file:// URI?" like so:
+
+```typescript
+server.resource(
+   "file",
+   new ResourceTemplate("file://{path}", { list: undefined }),
+   async (uri, { path }) => ({
+     // do something with the file at path, e.g., read its contents
+     contents: [{
+       uri: uri.href,
+       text: `File at ${path}`
+     }]
+   })
+ );
+```
+
+**Time beetle**: "Yes, exactly! The resource can be anything that provides data or functionality to the MCP server. You can define resources for files, databases, APIs, or any other data source you want to use."
+
+Now, let's add a transport layer to the server so it can communicate with clients."
+
+```typescript
+// Start receiving messages on stdin and sending messages on stdout
+const transport = new StdioServerTransport();
+await server.connect(transport); 
+```
+
+In this code, we:
+
+- Created a transport layer using the `StdioServerTransport` class, which allows the server to communicate with clients through standard input and output.
```

Comment:
This may be covered later, but if not you might differentiate between how stdio runs locally whereas other options such as SSE or streaming HTTP run remotely on a server.